In [1]:
import pandas as pd
from sklearn.tree import DecisionTreeRegressor

In [2]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import KFold
from sklearn.model_selection import train_test_split

In [3]:
import numpy as np

---

# 학습 목표

DecisionTreeClassifier 알고리즘의 정확도를 검증하기 위해 아래 검증 방법을 사용해보자.

- train_test_split()
- KFold()
- StratifiedKFold()

## 모델 인스턴스 생성

In [4]:
dt_clf = DecisionTreeClassifier()

## 데이터 load

In [5]:
iris = load_iris()
# iris_df = pd.DataFrame(data=iris.data, columns=iris.feature_names)
# iris_df['label']=iris.target
# iris_df.head()

## feature & label 분리

In [32]:
# feature = iris_df[iris.feature_names]
# label = iris_df['label']
feature = iris.data
label = iris.target

print(feature.shape)
print(label.shape)

(150, 4)
(150,)


# train_test_split()

<img src="https://builtin.com/sites/www.builtin.com/files/styles/ckeditor_optimize/public/inline-images/1_train-test-split_0.jpg"  width="800">

## 학습 세트와 검증 세트로 분리

In [7]:
X_train, X_test, y_train, y_test = train_test_split(feature, label,test_size=0.3, random_state=121)

## 모델 학습

In [8]:
dt_clf.fit(X_train, y_train)

DecisionTreeClassifier()

## 예측

In [42]:
pred = dt_clf.predict(X_test)
print('예측 정확도: {0:.4f}'.format(accuracy_score(y_test,pred)))

예측 정확도: 0.7667


---

# KFold()

데이터의 size는 150개이므로 5개로 나누면 테스트 데이터셋은 30개가 된다.

In [21]:
kfold = KFold(n_splits=5)
kfold

KFold(n_splits=5, random_state=None, shuffle=False)

In [41]:
lst_accuracy = []
n_iter = 0
 
for train_index, test_index in kfold.split(feature):
    X_train, X_test = feature[train_index], feature[test_index]
    y_train, y_test = label[train_index], label[test_index]

    dt_clf.fit(X_train , y_train)    
    pred = dt_clf.predict(X_test)
    n_iter += 1

    accuracy = np.round(accuracy_score(y_test,pred), 4)
    
    print('#{0} 검증 세트의 인덱스 범위:{1} ~ {2}'.format(n_iter,test_index[0],test_index[-1]))
    
    lst_accuracy.append(accuracy)

#1 검증 세트의 인덱스 범위:0 ~ 29
#2 검증 세트의 인덱스 범위:30 ~ 59
#3 검증 세트의 인덱스 범위:60 ~ 89
#4 검증 세트의 인덱스 범위:90 ~ 119
#5 검증 세트의 인덱스 범위:120 ~ 149


In [44]:
print(lst_accuracy)
print(np.mean(lst_accuracy))

[1.0, 1.0, 0.8333, 0.9333, 0.7667]
0.9066599999999999
